### ISSUE TITLE
请问oneflow有TripletLoss损失函数 和学习率设置相关的函数么

### ISSUE THREAD

**Issue by kanagi2587** [Oneflow-Inc/oneflow#3854]

**Comment by MARD1NO** [comment]
Triplet loss 可先参考以下代码
```
import oneflow as flow
import numpy as np
import oneflow.typing as tp
import oneflow.math as math
​
​
class _TripletLoss():
    """Triplet loss with hard positive/negative mining.
    Reference:
        Hermans et al. In Defense of the Triplet Loss for Person Re-Identification. arXiv:1703.07737.
    Imported from `<https://github.com/Cysu/open-reid/blob/master/reid/loss/triplet.py>`_.
    Args:
        margin (float, optional): margin for triplet. Default is 0.3.
    """
​
    def __init__(self, margin=0.3):
        self.margin = margin
​
    def _MarginRankingLoss(self, input1, input2, target, reduction='mean'):
        if reduction == 'none':
            ret = flow.clip(
                math.add(self.margin, math.multiply(target, math.multiply(-1, math.subtract(input1, input2)))),
                min_value=0)
        else:
            ret = math.reduce_mean(flow.clip(
                math.add(self.margin, math.multiply(target, math.multiply(-1, math.subtract(input1, input2)))),
                min_value=0))
        return ret
​
    def forward(self, inputs, targets):
        """
        Args:
            inputs (torch.Tensor): feature matrix with shape (batch_size, feat_dim).
            targets (torch.LongTensor): ground truth labels with shape (num_classes).
        """
        n = inputs.shape[0]
        dist = math.reduce_sum(math.pow(inputs, flow.constant_like(inputs, 2, dtype=flow.float32)), axis=1)
​
        shape_tensor = flow.constant(value=0.0, dtype=flow.float32, shape=(n, n))
        dist = flow.broadcast_like(dist, like=shape_tensor, broadcast_axes=[1])
​
        dist = math.add(dist, flow.transpose(dist, perm=(1, 0), batch_axis_non_change=True))
​
        temp1 = math.multiply(-2, flow.matmul(inputs, flow.transpose(inputs, perm=(1, 0), batch_axis_non_change=True)))
        dist = math.add(dist, temp1)
        dist = math.sqrt(flow.clamp(dist, min_value=1e-12))
​
        mask = math.equal(flow.broadcast_like(targets, like=shape_tensor, broadcast_axes=[1]),
                          flow.transpose(flow.broadcast_like(targets, like=shape_tensor, broadcast_axes=[1]),
                                         perm=(1, 0), batch_axis_non_change=True))
        mask_rev = math.not_equal(flow.broadcast_like(targets, like=shape_tensor, broadcast_axes=[1]),
                                  flow.transpose(flow.broadcast_like(targets, like=shape_tensor, broadcast_axes=[1]),
                                                 perm=(1, 0), batch_axis_non_change=True))
​
        dist_ap, dist_an = [], []
        for i in range(n):
            temp_dist = flow.slice_v2(dist, [(i, i + 1, 1)])
            temp_mask = flow.slice_v2(mask, [(i, i + 1, 1)])
            temp_mask_rev = flow.slice_v2(mask_rev, [(i, i + 1, 1)])
            dist_ap.append(math.reduce_max(flow.gather_nd(temp_dist, flow.where(temp_mask))))
            dist_an.append(math.reduce_min(flow.gather_nd(temp_dist, flow.where(temp_mask_rev))))
        dist_ap = flow.concat(dist_ap)
​
        dist_an = flow.concat(dist_an)
​
        y = flow.ones_like(dist_an)
​
        return self._MarginRankingLoss(dist_an, dist_ap, y)
```

**Comment by MARD1NO** [comment]
如果是需要 `torch.TripletMarginLoss` 可以参考以下代码：
```
def triplet_margin_loss(
    anchor: remote_blob_util.BlobDef,
    positive: remote_blob_util.BlobDef,
    negative: remote_blob_util.BlobDef,
    margin: float = 0.0,
    p: float = 2.0, 
    eps: float=1e-6, 
    swap: bool=False, 
    reduction: str = "mean",
    name: Optional[str] = None,
) -> remote_blob_util.BlobDef:
    assert reduction in [
        "none",
        "mean",
        "sum",
    ], "{} is not a valid value for reduction, The reduction must be the one of `none`, `mean`, `sum`. ".format(
        reduction
    )

    assert swap == False, "For now we only support `swap=True`, OneFlow still have backward error in minimum"

    if name is None:
        name = "TripletMarginLoss_"

    def _p_norm(x, p=2.0, name="p_norm"): 
        r"""Compute the p-norm 

        The equation is: 

        .. math:: 

            out = \sqrt[P]{\sum_{i=0}^{n}(abs(x)^P)} 

        Args:
            x ([type]): [description]
            p ([type], optional): [description]. Defaults to 2..
        """
        # In order to avoid the `nan` case. 
        _abs_val = flow.math.abs(x, name=name+"_abs")
        
        if p == 2.0: 
            # Use Square to compute the p2-norm
            _norm = flow.math.square(_abs_val, name=name+"_square")
            _norm = flow.math.reduce_sum(_norm, axis=-1, name=name+"_sum")
            _norm_val = flow.math.sqrt(_norm, name=name+"_sqrt")
        else: 
            _p_constant = flow.constant_like(like=_abs_val, value=p, dtype=flow.float32, name=name+"_p_constant")
            _norm = flow.math.pow(_abs_val, _p_constant, name=name+"_pow1")
            _norm = flow.math.reduce_sum(_norm, axis=-1, name=name+"_sum")
            _p_reciprocal_constant = flow.constant_like(like=_norm, value=1.0/p, dtype=flow.float32, name=name+"_p_reciprocal_constant")
            _norm_val = flow.math.pow(_norm, _p_reciprocal_constant, name=name+"_norm_val")
            
        return _norm_val

    # Compute the distance

    _distance_1 = _p_norm(anchor - positive + eps, p=p, name=name+"_distance_1")
    _distance_2 = _p_norm(anchor - negative + eps, p=p, name=name+"_distance_2")

    if swap:
        _distance_swap = _p_norm(positive - negative + eps, p=p)
        _distance_swap = flow.math.reduce_sum(_distance_swap, axis=-1)
        # TODO: minimum still not support backward
        _distance_2 = flow.math.minimum(_distance_2, _distance_swap)

    _triplet_loss = flow.clip(margin + _distance_1 - _distance_2, min_value=0.0)

    if reduction == "mean":
        return flow.math.reduce_mean(_triplet_loss, name=name+"_reduce_mean")
    elif reduction == "sum":
        return flow.math.reduce_sum(_triplet_loss, name=name+"_reduce_sum")
    else:
        return _triplet_loss
```

**Comment by kanagi2587** [comment]
> Reference



> Triplet loss 可先参考以下代码
> 
> ```
> import oneflow as flow
> import numpy as np
> import oneflow.typing as tp
> import oneflow.math as math
> ​
> ​
> class _TripletLoss():
>     """Triplet loss with hard positive/negative mining.
>     Reference:
>         Hermans et al. In Defense of the Triplet Loss for Person Re-Identification. arXiv:1703.07737.
>     Imported from `<https://github.com/Cysu/open-reid/blob/master/reid/loss/triplet.py>`_.
>     Args:
>         margin (float, optional): margin for triplet. Default is 0.3.
>     """
> ​
>     def __init__(self, margin=0.3):
>         self.margin = margin
> ​
>     def _MarginRankingLoss(self, input1, input2, target, reduction='mean'):
>         if reduction == 'none':
>             ret = flow.clip(
>                 math.add(self.margin, math.multiply(target, math.multiply(-1, math.subtract(input1, input2)))),
>                 min_value=0)
>         else:
>             ret = math.reduce_mean(flow.clip(
>                 math.add(self.margin, math.multiply(target, math.multiply(-1, math.subtract(input1, input2)))),
>                 min_value=0))
>         return ret
> ​
>     def forward(self, inputs, targets):
>         """
>         Args:
>             inputs (torch.Tensor): feature matrix with shape (batch_size, feat_dim).
>             targets (torch.LongTensor): ground truth labels with shape (num_classes).
>         """
>         n = inputs.shape[0]
>         dist = math.reduce_sum(math.pow(inputs, flow.constant_like(inputs, 2, dtype=flow.float32)), axis=1)
> ​
>         shape_tensor = flow.constant(value=0.0, dtype=flow.float32, shape=(n, n))
>         dist = flow.broadcast_like(dist, like=shape_tensor, broadcast_axes=[1])
> ​
>         dist = math.add(dist, flow.transpose(dist, perm=(1, 0), batch_axis_non_change=True))
> ​
>         temp1 = math.multiply(-2, flow.matmul(inputs, flow.transpose(inputs, perm=(1, 0), batch_axis_non_change=True)))
>         dist = math.add(dist, temp1)
>         dist = math.sqrt(flow.clamp(dist, min_value=1e-12))
> ​
>         mask = math.equal(flow.broadcast_like(targets, like=shape_tensor, broadcast_axes=[1]),
>                           flow.transpose(flow.broadcast_like(targets, like=shape_tensor, broadcast_axes=[1]),
>                                          perm=(1, 0), batch_axis_non_change=True))
>         mask_rev = math.not_equal(flow.broadcast_like(targets, like=shape_tensor, broadcast_axes=[1]),
>                                   flow.transpose(flow.broadcast_like(targets, like=shape_tensor, broadcast_axes=[1]),
>                                                  perm=(1, 0), batch_axis_non_change=True))
> ​
>         dist_ap, dist_an = [], []
>         for i in range(n):
>             temp_dist = flow.slice_v2(dist, [(i, i + 1, 1)])
>             temp_mask = flow.slice_v2(mask, [(i, i + 1, 1)])
>             temp_mask_rev = flow.slice_v2(mask_rev, [(i, i + 1, 1)])
>             dist_ap.append(math.reduce_max(flow.gather_nd(temp_dist, flow.where(temp_mask))))
>             dist_an.append(math.reduce_min(flow.gather_nd(temp_dist, flow.where(temp_mask_rev))))
>         dist_ap = flow.concat(dist_ap)
> ​
>         dist_an = flow.concat(dist_an)
> ​
>         y = flow.ones_like(dist_an)
> ​
>         return self._MarginRankingLoss(dist_an, dist_ap, y)
> ```
> 
> 这个是挖掘hard_sample的代码
> 
> 学习率相关函数可参考文档 [oneflow.optimizer](https://oneflow-api.readthedocs.io/en/latest/optimizer.html)
> 
> 包含了常见的 Cosinedecay, Exponentialdecay等等衰减函数

请问该损失函数能够正常运作么，因为我发现我所需要的函数的Reference和这个指向是一样的，但是测试过程中发现得到的loss总是为0

**Comment by MARD1NO** [comment]
这个损失函数是正常运作的，之江实验室中的项目里面采取的就是这部分代码，来做难样本挖掘的损失函数。

如果没有正常运作，能否贴一下你的测试案例呢？

另外，新版本已经合并了与torch.TripletMarginLoss相同的损失函数，可参考 #3864 该PR，里面的注释有十分详细的使用案例

**Comment by kanagi2587** [comment]
> 这个损失函数是正常运作的，之江实验室中的项目里面采取的就是这部分代码，来做难样本挖掘的损失函数。
> 
> 如果没有正常运作，能否贴一下你的测试案例呢？
> 
> 另外，新版本已经合并了与torch.TripletMarginLoss相同的损失函数，可参考 #3864 该PR，里面的注释有十分详细的使用案例
似乎是dist_an, dist_ap, y的量级差得太多导致发生了这个情况，不知道是什么导致的
[3.4028235e+38 3.4028235e+38 3.4028235e+38 3.4028235e+38]
[0.06074639 0.07050213 0.07050213 0.06081993]
[1. 1. 1. 1.]

**Comment by MARD1NO** [comment]
> > 这个损失函数是正常运作的，之江实验室中的项目里面采取的就是这部分代码，来做难样本挖掘的损失函数。
> > 如果没有正常运作，能否贴一下你的测试案例呢？
> > 另外，新版本已经合并了与torch.TripletMarginLoss相同的损失函数，可参考 #3864 该PR，里面的注释有十分详细的使用案例
> > 似乎是dist_an, dist_ap, y的量级差得太多导致发生了这个情况，不知道是什么导致的
> > [3.4028235e+38 3.4028235e+38 3.4028235e+38 3.4028235e+38]
> > [0.06074639 0.07050213 0.07050213 0.06081993]
> > [1. 1. 1. 1.]

那应该检查你输入进去的数值是不是过大，不合法了。
